ocaml/xapi/stream_vdi.ml

```diff
@@ -100,6 +103,98 @@ let write_block ~__context filename buffer ofd len =
+  (major, minor)
+
+let is_nbd_device path =
+  let nbd_device_num = 43 in
+  let (major, _) = get_device_numbers path in
+  major = nbd_device_num
+
+type nbd_connect_info =
+  { path : string
+  ; exportname : string
+  } [@@deriving rpc]
+
+let get_nbd_device path =
+  let nbd_device_prefix = "/dev/nbd" in
+  if Astring.String.is_prefix ~affix:nbd_device_prefix path && is_nbd_device path then begin
+    let nbd_number =
+      String.sub path (String.length nbd_device_prefix) (String.length path - String.length nbd_device_prefix)
+    in
+    let { path; exportname } =
+      let persistent_nbd_info_dir = "/var/run/nonpersistent/nbd" in
+      let filename = persistent_nbd_info_dir ^ "/" ^ nbd_number in
+      Xapi_stdext_unix.Unixext.string_of_file filename
+      |> Jsonrpc.of_string
+      |> nbd_connect_info_of_rpc
+    in
+    Some (path, exportname)
+  end else None
+
+type extent = {
+  flags : int32;
+  length : int64;
+} [@@deriving rpc]
+
+type extent_list = extent list [@@deriving rpc]
+
+(* Flags for extents returned for the base:allocation NBD metadata context, documented at https://github.com/NetworkBlockDevice/nbd/blob/master/doc/proto.md *)
+let flag_hole = 1l
+let flag_zero = 2l
+
+(** [descriptor_list] should be an list of non-overlapping extents, ordered from lowest offset to highest. *)
+let cycle_descriptors descriptor_list offset =
+  (* Output range includes start and end points *)
+  let rec range acc start_chunk end_chunk =
+    if end_chunk < start_chunk then acc
+    else range (end_chunk::acc) start_chunk Int64.(add end_chunk minus_one)
+  in
+
+  let is_empty e =
+    let has_flag flag =
+      Int32.logand e.flags flag = flag
+    in
+    (* We assume the destination is prezeroed, so we do not have to copy zeroed extents *)
+    (has_flag flag_hole) || (has_flag flag_zero)
+  in
+
+  let find_blocks descriptor offset increment =
+    match (is_empty descriptor) with
+      | false -> begin
+        let start_chunk = Int64.div offset increment in
+        (* If the chunk ends at the boundary don't include the next chunk *)
+        let end_chunk = let open Int64 in div (add (add descriptor.length offset) minus_one) increment in
+        let chunks = range [] start_chunk end_chunk in
+        chunks, descriptor.length
+        end
+      | true -> [], descriptor.length
+  in
+  (* Only compare to most recent chunk added*)
+  let add_new a b =
+    match b with
+    | hd :: _ when Int64.equal a hd -> b
+    | _ -> a::b
+  in
+
+  let rec add_unique_chunks acc = function
+    | [] -> List.rev acc
+    | x::xs -> add_unique_chunks (add_new x acc) xs
+  in
+
+  let rec process acc offset = function
+    | [] -> acc, offset
+    | x::xs -> begin
+      let chunks, add_offset = find_blocks x offset chunk_size in
+      process (add_unique_chunks (List.rev acc) chunks) (Int64.add offset add_offset) xs
+      end
+  in
+
+  let chunks, _ = process [] offset descriptor_list  in
```

Comment:
Is the intent here to filter out zeroes and holes? Could we simply use a List.filter is_empty_descriptor instead?

Reply:
There is, but I think it's not worth making so big a change now, because this code will be replaced eventually.

Reply:
Should I leave this as is than?

Reply:
I think @gaborigloi mentioned a problem that this is currently quadratic, which could make this slow. If filtering is what you actually want to do, and your list of offsets is already sorted it might be easier to do that instead of building a lot of intermediate lists, it would simplify the code and make it smaller.

Reply:
> It is, and then we split them into chunks, but I think it's not worth making so big a change now, because this code will be replaced eventually.

What will it get replaced with? Could we implement the replacement directly?

Reply:
we have already removed the quadratic part in last commit. Regarding changing this code, when every SR type will support CBT API we can use the same to query sparseness but we can not do that now.
@edwintorok Are you OK wit it? If not, can you tell me how should I modify the existing code?